### ISSUE TITLE
Dynamic Keyboard Enhancement - Size, Position, and Docking

### ISSUE THREAD

**Issue by AdamRoden** [OptiKey/OptiKey#588]
In the current XML, you can define a keyboard's height as a percentage of the screen. If omitted the size does not change from the previous value. 

This enhancement is to expand on that logic and incorporate the following:
1. Width expressed as either an absolute value or percentage of screen. Examples: [1920|100%]
2. Height expressed as either an absolute value or percentage of screen. Examples: [360|30%]
3. X Position expressed as either an absolute value or percentage of screen. Examples: [1760|90%]
4.  Y Position expressed as either an absolute value or percentage of screen. Examples: [0|70%]
5. Docked Location. Examples: [Undock|Left|Right|Top|Bottom]

**Comment by JuliusSweetland** [comment]
@AdamRoden Understood. Let's keep in touch as you implement this so we can discuss the placement parameters as you go.

Regarding 3: I don't believe eye trackers typically output any co-ordinates outside of the calibrated screen dimensions, so this may not be possible.

**Comment by AdamRoden** [comment]
@JuliusSweetland After looking into the code, I've found some pieces to re-use, like the enums for MoveToDirections and WindowStates, so I will base my XML values off of those.

Regarding 3:  If true, that is a bit disappointing. I only know that it has worked for me with my Tobii 4C, even in OptiKey. If I use another program to force my keyboard outside my screen, OptiKey still activates the correct off screen keys.

**Comment by JuliusSweetland** [comment]
Ah then from your experience it sounds like I am wrong. Ignore me.

**Comment by AdamRoden** [comment]
Julius,
The code is starting to come together quite nicely now.
At this point I'm testing my new code for xml elements tentatively named
'WindowState', 'MoveToDirection', 'Height', 'Width', 'HorizontalOffset',
and 'VerticalOffset. At this point, valid element combinations are as
follows:

   1. WindowState = Maximised
   2. WindowState = Docked
      - MoveToDirection = [Top|Bottom]
         - Height = [nnn% | nnnn]
      - MoveToDirection = [Left|Right]
         - Width = [nnn% | nnnn]
      3. WindowState = Floating
      -  MoveToDirection =
      [Top|TopLeft|TopRight|Left|Right|Bottom|BottomLeft|BottomRight]
      - Height = [nnn% | nnnn]
      - Width = [nnn% | nnnn]
      - HorizontalOffset  = [nnn% | nnnn] (negative implies offset leftward)
      - VerticalOffset = [nnn% | nnnn] (negative implies offset upward)

Let me know if you have any more thoughts or concerns on this. Otherwise,
if what I've laid out makes sense and if you don't have any additional
proposals, I will move forward and finish coding and testing.

Lastly, with what I've learned here, the next enhancement I work on could
be the new xml function keys to get and set values. I don't want to step on
your toes, so think about it and let me know.

Adam Roden

**Comment by JuliusSweetland** [comment]
This all looks good. Could you expand on 3 a bit more. I presume the user
would select a key which would change OptiKey to be floating, but then what
would it mean for the key to also move to top by x (for example). The
movement is relative to what?

**Comment by AdamRoden** [comment]
@JuliusSweetland Good questions. I guess using "MoveToDirection" as the label might be confusing when really it is more like where to set the origin point. Conceptually, what would happen is the top center of the keyboard would jump to the top center of the screen as if it were being docked, but we are not docking. Then, the offset values get applied to set the final position of the keyboard.

**Comment by JuliusSweetland** [comment]
Ah that makes sense. Thanks for clarifying.

**Comment by tqphan** [comment]
I have 2 custom keyboards 

keyboard 1:
```
<Keyboard>
    <Name>MyResize 25%</Name>
    <HideFromKeyboardMenu>False</HideFromKeyboardMenu>
    <ShowOutputPanel>False</ShowOutputPanel>
	<!--PersistNewState: if true this keyboard's state, size, and position settings will become the default values to carry forward when changing keyboards-->
	<!--if element is not included or not true this keyboard's settings are only applicable once, and will not carry forward-->
	<PersistNewState>False</PersistNewState>  
	<!--WindowState: Floating, Docked-->
	<!--if element is not included or does not contain a value from the list then use the saved default value-->
	<WindowState>Docked</WindowState>
	<!--Position: TopLeft, Top, TopRight, Left, Right, BottomLeft, Bottom, BottomRight-->
	<!--if element is not included or does not contain a value from the list then use the saved default value-->
	<Position>Bottom</Position> 
	<!--Width and Height: -9999 to 9999. % implies percent of screen. if negative then use the sum of the value and the screen's dimension-->
	<!--i.e. if a Width of -300 is sent and the screen is 1920 wide then the keyboard width will be set to 1620 by way of ((-300) + (1920) = (1620))-->
	<Height>25%</Height>
    <SymbolMargin>4</SymbolMargin>
    <BorderThickness>1,1,1,1</BorderThickness>
```

Keyboard 2:

```
<Keyboard>
    <Name>MyResize 25%</Name>
    <HideFromKeyboardMenu>False</HideFromKeyboardMenu>
    <ShowOutputPanel>False</ShowOutputPanel>
	<!--PersistNewState: if true this keyboard's state, size, and position settings will become the default values to carry forward when changing keyboards-->
	<!--if element is not included or not true this keyboard's settings are only applicable once, and will not carry forward-->
	<PersistNewState>False</PersistNewState>  
	<!--WindowState: Floating, Docked-->
	<!--if element is not included or does not contain a value from the list then use the saved default value-->
	<WindowState>Docked</WindowState>
	<!--Position: TopLeft, Top, TopRight, Left, Right, BottomLeft, Bottom, BottomRight-->
	<!--if element is not included or does not contain a value from the list then use the saved default value-->
	<Position>Bottom</Position> 
	<!--Width and Height: -9999 to 9999. % implies percent of screen. if negative then use the sum of the value and the screen's dimension-->
	<!--i.e. if a Width of -300 is sent and the screen is 1920 wide then the keyboard width will be set to 1620 by way of ((-300) + (1920) = (1620))-->
	<Height>50%</Height>
    <SymbolMargin>4</SymbolMargin>
    <BorderThickness>1,1,1,1</BorderThickness>
```

The only difference is `<Height>25%</Height>` and `<Height>50%</Height>`

This is what I'm experiencing: https://youtu.be/qP3WlLFmTbo

**Comment by JuliusSweetland** [comment]
@AdamRoden please see the video in the previous comment. It looks like OptiKey is resizing over and over for some reason and also persisting the size changes so they impact the rest of OptiKey. One mechanism that may be coming into play here is that when you change a dock height Windows can call back to OptiKey with the new height, which can be changed by Windows. I'm not sure that's what's happening here, but the flickering has historically been connected to that

**Comment by AdamRoden** [comment]
@tqphan @JuliusSweetland Thanks for the feedback! That screen flickering was bad. Julius is on point regarding Windows and docking being related to the flickering. I did a deep dive into the code today and got a stronger understanding of what's going on, and patched a lot of thing, and I now have much cleaner code on my PC, but need to review it with Julius before it can be pulled in.

**Comment by AdamRoden** [comment]
Most of the issues are resolved in release v2.30.0, but there is a problem if you try to use a dynamic keyboard with DockSize of Collapsed. Fixing it is on my to do list.

**Comment by JuliusSweetland** [comment]
Great news. Thanks Adam
